service.go

```diff
@@ -0,0 +1,249 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      operatorName,
+			Namespace: namespace,
+			Labels:    label,
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Service",
+			APIVersion: "v1",
+		},
+		Spec: v1.ServiceSpec{
+			Ports: []v1.ServicePort{
+				{
+					Port:     port,
+					Protocol: v1.ProtocolTCP,
+					TargetPort: intstr.IntOrString{
+						Type:   intstr.Int,
+						IntVal: port,
+					},
+					Name: portName,
+				},
+			},
+			Selector: label,
+		},
+	}
+
+	return service, nil
+}
+
+// GenerateServiceMonitor generates a prometheus-operator ServiceMonitor object
+// based on the passed Service object.
+func GenerateServiceMonitor(s *v1.Service) *monitoringv1.ServiceMonitor {
+	labels := make(map[string]string)
+	for k, v := range s.ObjectMeta.Labels {
+		labels[k] = v
+	}
+
+	return &monitoringv1.ServiceMonitor{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ServiceMonitor",
+			APIVersion: "monitoring.coreos.com/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      s.ObjectMeta.Name,
+			Namespace: s.ObjectMeta.Namespace,
+			Labels:    labels,
+		},
+		Spec: monitoringv1.ServiceMonitorSpec{
+			Selector: metav1.LabelSelector{
+				MatchLabels: labels,
+			},
+			Endpoints: []monitoringv1.Endpoint{
+				{
+					Port: s.Spec.Ports[0].Name,
+				},
+			},
+		},
+	}
+}
+
+// Create route to the specified service.
+func GenerateRoute(s *v1.Service, path string) *routev1.Route {
+	labels := make(map[string]string)
+	for k, v := range s.ObjectMeta.Labels {
+		labels[k] = v
+	}
+
+	return &routev1.Route{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Route",
+			APIVersion: "route.openshift.io/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      s.ObjectMeta.Name,
+			Namespace: s.ObjectMeta.Namespace,
+			Labels:    labels,
+		},
+		Spec: routev1.RouteSpec{
+			Path: path,
+			To: routev1.RouteTargetReference{
+				Kind: "Service",
+				Name: s.ObjectMeta.Name,
+			},
+			Port: &routev1.RoutePort{
+				TargetPort: s.Spec.Ports[0].TargetPort,
+			},
+		},
+	}
+}
+
+// ConfigureMetrics generates metrics service and route,
+// creates the metrics service and route,
+// and finally it starts the metrics server
+func ConfigureMetrics(ctx context.Context, metricpath string, list []prometheus.Collector, fn recordmetrics) error {
+	log.Info("Starting prometheus metrics")
+	StartMetrics(list, fn)
+
+	client, err := createClient()
+	if err != nil {
+		log.Info("Failed to create new client", "Error", err.Error())
+		return nil
+	}
+
+	// Generate Service Object
+	s, svcerr := GenerateService(8080, "metrics")
+	if svcerr != nil {
+		log.Info("Error generating metrics service object.", "Error", svcerr.Error())
+		return ErrMetricsFailedGenerateService
+	}
+	log.Info("Generated metrics service object")
+
+	// Create or update Service
+	_, err = createOrUpdateService(ctx, client, s)
+	if err != nil {
+		log.Info("Error getting current metrics service", "Error", err.Error())
+		return ErrMetricsFailedCreateService
```

Comment:
I usually find the actual error more useful than a custom error.

Reply:
This was part of some go client stuff @rafael-azevedo was working on.

Reply:
Modified to return actual error.

Reply:
I am unfamiliar with the go client work the @rafael-azevedo was working on. Do we want to reuse these errors?

Reply:
I think its just a convention. I find it helpful but its definitely a debatable thing to keep.  @rafael-azevedo ?